python/hsfs/core/tfdata_engine.py

```diff
@@ -0,0 +1,461 @@
+#
+#   Copyright 2020 Logical Clocks AB
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+import re
+import itertools
+
+try:
+    import tensorflow as tf
+except ModuleNotFoundError:
+    pass
+
+try:
+    from pydoop import hdfs
+except ModuleNotFoundError:
+    pass
+
+try:
+    import boto3
+except ModuleNotFoundError:
+    pass
+
+
+class TFDataEngine:
+    SUPPORTED_TFDTYPES = [
+        tf.float16,
+        tf.float32,
+        tf.float64,
+        tf.int16,
+        tf.int64,
+        tf.int32,
+    ]
+    SPARK_TO_TFDTYPES_MAPPINGS = {
+        "string": tf.string,
+        "short": tf.int16,
+        "int": tf.int32,
+        "long": tf.int64,
+        "float": tf.float32,
+        "double": tf.float64,
+    }
+
+    def __init__(
+        self,
+        training_dataset,
+        split,
+        target_name,
+        feature_names,
+        is_training,
+        cycle_length,
+    ):
+
+        self._training_dataset = training_dataset
+        self._split = split
+        self._target_name = target_name
+        self._feature_names = feature_names
+        self._is_training = is_training
+        self._cycle_length = cycle_length
+
+        self._features = self._training_dataset._features
+        self._training_dataset_format = self._training_dataset.data_format
+
+        self._input_files = _get_training_dataset_files(
+            self._training_dataset.location,
+            self._split,
+            filter_empty=True if self._training_dataset_format == "csv" else False,
+        )
+
+        if self._feature_names is None:
+            self._feature_names = [feat.name for feat in self._features]
+
+            if target_name in self._feature_names:
+                self._feature_names.remove(target_name)
+
+    def get_serialized_example_schema(self):
+        """
+        returns schema for parsing serialized example of ff record file
+        :return: schema for tfrecord file
+        :rtype: dict
+        """
+        _, tfrecord_feature_description = _get_tfdataset(
+            self._input_files, self._cycle_length
+        )
+        return tfrecord_feature_description
+
+    def tf_record_dataset(
+        self,
+        batch_size=None,
+        num_epochs=None,
+        one_hot_encode_labels=False,
+        num_classes=None,
+        process=False,
+        serialized_ndarray_fname=[],
+    ):
```

Comment:
you should add the types here as python type annotations https://www.python.org/dev/peps/pep-0484/